src/server/journal/streamer.h

```diff
@@ -60,14 +59,25 @@ class JournalStreamer : public journal::JournalConsumerInterface {
   ExecutionState* cntx_;
 
  private:
-  void AsyncWrite();
-  void OnCompletion(std::error_code ec, size_t len);
-
   bool IsStalled() const;
 
   journal::Journal* journal_;
 
-  PendingBuf pending_buf_;
+  util::fb2::Fiber periodic_writer_;
+  util::fb2::Done periodic_writer_done_;
+  void PeriodicWriterFiber(std::chrono::microseconds period_ms, util::fb2::Done* waiter);
+  void SendAsync(bool writer_done);
+  void OnCompletion(std::error_code ec, size_t len, bool writer_done);
+
+  using JournalStreamerItem = struct JournalStreamerItem {
+    size_t size;
+    std::string data;
+  };
+
+  static constexpr uint32_t in_flight_data_size_limit = 1024;
+  moodycamel::ReaderWriterQueue<JournalStreamerItem> queue_;
```

Comment:
Assuming all producers and the single consumer run on the same thread (each of them on a separate fiber) why do we need the `atomics` ? If everything runs on the same thread concurrently then we do not need any synchronization primitive (unless we have partial writes from the producer to the consumer that must be satisfied fully -- upon on that case a counting barrier without atomic will suffice)

Reply:
I know it's not ready for review or anything, just saw this and thought it would be worth commenting

Reply:
As @kostasrim suggested - moving to use non atomic container.